relay-server/src/utils/rate_limits.rs

```diff
@@ -378,18 +387,59 @@ where
         let mut enforcement = Enforcement::default();
 
         if let Some(category) = summary.event_category {
-            let event_limits = (self.check)(scoping.item(category), 1)?;
-            let longest = event_limits.longest();
-            enforcement.event = CategoryLimit::new(category, 1, longest);
-
-            // Record the same reason for attachments, if there are any.
-            enforcement.attachments = CategoryLimit::new(
-                DataCategory::Attachment,
-                summary.attachment_quantity,
-                longest,
-            );
-
-            rate_limits.merge(event_limits);
+            if category == DataCategory::Transaction {
+                if summary.transaction_metrics_extracted {
+                    // Check processing quota, but DO NOT increment.
+                    let processing_limits =
+                        (self.check)(scoping.item(DataCategory::TransactionProcessed), 0)?;
+                    let longest = processing_limits.longest();
+
+                    // Reject the entire transaction event as we do not want to index it nor
+                    // extract metrics from it.  Record TransactionProcessed as 1, since
+                    // exactly one extraction is being denied, even if we didn't increment
+                    // the counter.  This works around a quirk of CategoryLimit::is_active
+                    // which only considers something limited if the count is larger than 1.
+                    enforcement.event = CategoryLimit::new(DataCategory::Transaction, 1, longest);
+                    enforcement.extracted_transaction_metrics =
+                        CategoryLimit::new(DataCategory::TransactionProcessed, 1, longest);
+                    rate_limits.merge(processing_limits);
+
+                    if !enforcement.event.is_active() {
+                        // Check and increment indexing quota.
+                        let index_limits =
+                            (self.check)(scoping.item(DataCategory::Transaction), 1)?;
+                        let longest = index_limits.longest();
+                        enforcement.event =
+                            CategoryLimit::new(DataCategory::Transaction, 1, longest);
+                        rate_limits.merge(index_limits);
+                    }
+                } else {
```

Comment:
Unless I'm missing something, both branches look identical with the one difference that you run another check for the event. If that's the case, then this could be rewritten with less code duplication.

Reply:
yup, I kept the control flow literally as in the design doc to start with.  Was expecting maybe tighten this down during review once we're sure it's the right logic.

Reply:
Makes sense, thank you.